src/details/ArborX_DetailsBatchedQueries.hpp

```diff
@@ -47,32 +47,34 @@ struct BatchedQueries
   // indirection when recording results rather than using that function at
   // the end.  We decided to keep reversePermutation around for now.
 
-  template <typename ExecutionSpace, typename Predicates>
+  template <typename ExecutionSpace, typename Predicates,
+            typename SpaceFillingCurve>
   static Kokkos::View<unsigned int *, DeviceType>
-  sortQueriesAlongZOrderCurve(ExecutionSpace const &space,
-                              Box const &scene_bounding_box,
-                              Predicates const &predicates)
+  sortPredicateAlongSpaceFillingCurve(ExecutionSpace const &space,
+                                      SpaceFillingCurve const &curve,
+                                      Box const &scene_bounding_box,
+                                      Predicates const &predicates)
   {
     using Access = AccessTraits<Predicates, PredicatesTag>;
     auto const n_queries = Access::size(predicates);
 
-    Kokkos::View<unsigned int *, DeviceType> morton_codes(
+    using OrderValueType =
+        KokkosExt::detected_t<SpaceFillingCurveProjectionArchetypeExpression,
+                              SpaceFillingCurve, Point>;
+    Kokkos::View<OrderValueType *, DeviceType> projections(
         Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
-                           "ArborX::BVH::query::morton"),
+                           "ArborX::BVH::query::projections"),
         n_queries);
     Kokkos::parallel_for(
-        "ArborX::BatchedQueries::assign_morton_codes_to_queries",
+        "ArborX::BatchedQueries::project_predicates_onto_space_filling_curve",
         Kokkos::RangePolicy<ExecutionSpace>(space, 0, n_queries),
         KOKKOS_LAMBDA(int i) {
-          using Details::returnCentroid;
-          Point xyz = returnCentroid(getGeometry(Access::get(predicates, i)));
-          translateAndScale(xyz, xyz, scene_bounding_box);
-          // Use 32-bit Morton indices instead of 64-bit as in construction. For
-          // most (all?) situations, 64-bit just adds a penalty with no benefit.
-          morton_codes(i) = morton32(xyz[0], xyz[1], xyz[2]);
+          projections(i) =
+              curve(scene_bounding_box,
+                    returnCentroid(getGeometry(Access::get(predicates, i))));
```

Comment:
You don't need to use `returnCentroid` here, I think, now that you can dispatch to different functions in `Morton32`.